ENH: LHS based OptimalDesign (scipy.stats.qmc)

tupui opened the pull request:
This adds a new QMC engine base on optimized LHS.

This was removed from `scipy.stats.qmc` during its review  (#10844).

The optimization loop consists in randomly permuting some points from the design. LHS properties are preserved doing so. All design candidates are compared in terms of centered-C2 discrepancy. This is commonly done in the literature. Also, recomputing the C2 discrepancy is helped with a special method which just update the discrepancy (`_perturb_disrepancy` from Jin et al. corrected as there where mistakes in the article).

The optimization consists in two loops. An outer loop and an inner loop. For more details, see the reference article: `Jin et al., "An efficient algorithm for constructing optimal design of computer experiments." Journal of Statistical Planning and Inference, 2005.`

Here is the flowchart of the optimization

<img width="521" alt="Screenshot 2021-04-18 at 18 58 54" src="https://user-images.githubusercontent.com/23188539/115153822-2a3b9680-a078-11eb-8241-54fc998fbfe4.png">

mdhaber commented:
Does it need to be updated after the LHS bug fix?

tupui commented:
>  Does it need to be updated after the LHS bug fix?

I will double check (and fix the conflict), but I updated it after.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -383,6 +385,98 @@ def update_discrepancy(x_new, sample, initial_disc):
 
 
+def _perturb_discrepancy(sample, i1, i2, k, disc):
+    """Centered discrepancy after and elementary perturbation on a LHS.
```
```suggestion
    """Centered discrepancy after an elementary perturbation on a LHS.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -383,6 +385,98 @@ def update_discrepancy(x_new, sample, initial_disc):
+    Returns
+    -------
+    discrepancy : float
+        Centered discrepancy.
```
```suggestion
        Centered discrepancy of the design after permutation.
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+    Generate samples from an optimal design.
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.OptimalDesign(d=2, seed=12345)
```
I have to update this.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+
+    """
+
+    def __init__(self, d, start_design=None, niter=1, method=None, seed=None):
```
Maybe I should add another option to change the metric. Some people might want to use the Linf or another distance. It would be completely generic then (with specialized code for C2). Speaking of C2, I will test it but I might have to put this in Cython to see any benefit since we have cython implementations of all discrepancies.

mdhaber replied:
Not yet. Let's start with the basic algorithm and documentation.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+        if n == 1:
+            return self.best_doe
+
+        def _perturb_best_doe(x: np.ndarray) -> float:
```
Also this could be a generic perturbation method.

Maybe I should have a base class for optimizing designs, and subclass to propose this one. Then a user could have his own class. It would follow what we have with `QMCEngine`.

mdhaber replied:
Same thought as above, please. Your optimizer doesn't do everything at once either; it iterates! : )

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -383,6 +385,98 @@ def update_discrepancy(x_new, sample, initial_disc):
+    z_ij = sample - 0.5
+
+    # Eq (19)
+    c_i1j = 1. / n ** 2. * np.prod(0.5 * (2. + abs(z_ij[i1, :]) +
+                                          abs(z_ij) -
+                                          abs(z_ij[i1, :] - z_ij)),
```
PEP8 now prefers operator at the beginning of a line. 
I would perefer to start parentheses right after the `=` sign so that this fits on two lines.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -383,6 +385,98 @@ def update_discrepancy(x_new, sample, initial_disc):
     return initial_disc + disc1 + disc2 + disc3
 
 
+def _perturb_discrepancy(sample, i1, i2, k, disc):
```
This function does seem accurate, and it is faster for large designs.

tupui replied:
Thanks for confirming this point. I will still see if I can do some cython.

mdhaber replied:
Sounds like Pythran is the way to go now.

mdhaber reviewed (changes requested):
Interesting, and seems very useful. I didn't check the implementation of `_perturb_discrepancy` but it seems to work well in my own test, which I included here. Some documentation suggestions, but probably more important are the questions about the use of `basinhopping`. Does it perform well on a function that has memory of past inputs? Does `"L-BFGS-B"` do well with its inputs being rounded?

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+    >>> qmc.discrepancy(sample_2)
+    0.01600398742881648
+
+    Finally, samples can be scaled to bounds.
+
+    >>> l_bounds = [0, 2]
+    >>> u_bounds = [10, 5]
+    >>> qmc.scale(sample, l_bounds, u_bounds)
+    array([[8.40526909, 3.609945  ],
+           [1.54532796, 3.08795949],
+           [5.2177809 , 4.80031164],
+           [6.80338249, 2.1942472 ],
+           [2.65448791, 3.83491828]])
+
```
Agreed. We can remove all (or most) of these examples.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+    Generate samples from an optimal design.
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.OptimalDesign(d=2, seed=12345)
```
Totally, I added a note to myself above. I did not update this branch since my adventures with seeds.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
 
 
+class OptimalDesign(QMCEngine):
+    """Optimal design.
```
You mean for the name of the class? `OptimalLatinHypercube` would be more appropriate I guess. And we could keep this class for the base class if we go as I suggested (https://github.com/scipy/scipy/pull/13471#discussion_r614576013) with a modular approach.

mdhaber replied:
I really meant for the first line of the docstring. But yes, `OptimizedLatinHypercube` might be better than `OptimalDesign` unless the latter always refers to a latin hypercube.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+    d : int
+        Dimension of the parameter space.
+    start_design : array_like (n, d), optional
+        Initial design of experiment to optimize. `OrthogonalLatinHypercube`
```
Indeed, it's a leftover.
Good point, I will add validations and tests. Thanks.

tupui commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -168,48 +169,66 @@ def test_update_discrepancy(self):
             update_discrepancy(x_new, space_1[:-1], disc_init)
 
+    def test_perm_discrepancy(self):
+        doe_init = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+        doe_init = (2.0 * doe_init - 1.0) / (2.0 * 6.0)
+
+        disc_init = qmc.discrepancy(doe_init)
+
+        row_1, row_2, col = 5, 2, 1
+
+        doe = doe_init.copy()
+        doe[row_1, col], doe[row_2, col] = doe[row_2, col], doe[row_1, col]
+
+        disc_valid = qmc.discrepancy(doe)
+        disc_perm = _perturb_discrepancy(doe_init, row_1, row_2, col,
+                                         disc_init)
+
+        assert_allclose(disc_valid, disc_perm)
```
Agreed I will update the tests.
Thanks for the analysis. It's indeed interesting to see the error increasing. I wonder what is causing this, some accumulation somewhere? As long as it stays small it's fine though.

mdhaber replied:
I guess I'm surprised that the error accumulates linearly; if it is not unstable, I would expect it to be more random, even at large scale. But I would be surprised if the equations are _wrong_ since it is so close

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+            """
+            # Perturb the DoE
+            doe = self.best_doe.copy()
+            col, row_1, row_2 = np.round(x).astype(int)
```
If I remember correctly, I coded this before SciPy 1.2 was out with `dual_annealing`. I will try to use it instead as it's what people use in the literature. Do you have another recommendation otherwise?

tupui replied:
Replied here https://github.com/scipy/scipy/pull/13471#discussion_r614633695

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+            self.best_disc = discrepancy(self.best_doe)
+
+        if n == 1:
+            return self.best_doe
```
TBH I don't remember why I did this. Maybe it's also a leftover. Thanks I will remove it as it seems useless otherwise.

mdhaber replied:
NM. `n` is the size of the sample, not number of iterations, so it makes sense

tupui replied:
Ah yes!

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+    d : int
+        Dimension of the parameter space.
+    start_design : array_like (n, d), optional
+        Initial design of experiment to optimize. `OrthogonalLatinHypercube`
```
```suggestion
        Initial design of experiment to optimize. `LatinHypercube`
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+        Initial design of experiment to optimize. `OrthogonalLatinHypercube`
+        is used to generate a first design otherwise.
+    niter : int, optional
+        Number of iterations to perform. Default is 1.
```
Correct, it is intended to be one execution of the optimizer.

mdhaber replied:
OK, something that explains that in the documentation would be helpful.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+        x0 = [0, 0, 0]
+        bounds = ([0, self.d - 1],
+                  [0, n - 1],
+                  [0, n - 1])
```
You are correct, the design keeps updating. See https://github.com/scipy/scipy/pull/13471#discussion_r614633695, `dual_annealing` will be better.

tupui commented:
Thanks for the review Matt. To summarise:

- [ ] I will change the optimizer to `dual_annealing`.
- [ ] Add validation and tests + your extended version for the update.
- [ ] Fix small PEP8 and things.
- [ ] Do I add a metric for the objective function? Ability to do Linf instead of C2 for instance.
- [ ] Making `_perturb_best_doe` a parameter? Not so sure as this would require extra logic. Or this could be done using a main optimization class. We would have `OptimalDesign(ABC)` and `OptimalLatinHypercube(OptimalDesign)`.

The metric is easy and valuable. The other one is more tricky. Optimizing a sample is already something advanced and optimized LHS is commonly used. For the rest it might be too custom so that we have a generic and useful interface.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,177 @@ def reset(self):
+
+            """
+            # Perturb the DoE
+            doe = self.best_doe.copy()
```
This is going to be really expensive and it shouldn't be needed because `_perturb_discrepancy` doesn't actually need to do the permutation to calculate the new discrepancy, right? Even if it did, it would be faster to un-permute (if thepermutation is not an improvement) than to copy everything every iteration.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,177 @@ def reset(self):
+        Dimension of the parameter space.
+    start_design : array_like (n, d), optional
+        Initial design of experiment to optimize. `LatinHypercube`
+        is used to generate a first design otherwise.
```
If a `start_design` is provided, all future calls to `random` will start from it, so they are not very distinct. This may be confusing.

tupui replied:
Indeed. I guess I will then just restart from a blank LHS?

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,188 @@ def reset(self):
+    niter : int, optional
+        Number of iterations to perform. Default is 1.
+    method : callable ``f(func, x0, bounds)``, optional
+        Optimization function used to search new samples. Default to
```
Currently, `basinhopping` does not appear to be performing really well; it seems to explore a very small portion of the possible swaps before terminating. The output design is very similar to the input after one iteration, and they differ only in small-number rows/cols. (e.g. use `np.where(before != after)`to see the differences).

Whatever optimizer you do end up with, I'd like to see it tested head-to-head against random permutations to see how the discrepancy goes down as a function of time for each. I'm not sure if the overhead of an optimizer is really going to be worth it, since the objective function keeps changing on them.

tupui replied:
This is great. Thanks for working hard on this! I will definitely use your script to test `dual_annealing`. I will also vary the dimension. Not sure you remember this, but the original code I had for the PR on QMC had 2 strategies. It was first checking the complexity of the problem (just calculating the total number of possible combinations). And if the complexity was bellow a threshold, it was going through all the possible permutations. Otherwise it would use basinhopping.
I am curious to see how `dual_annealing` performs here. Everyone uses it in the literature so I hope to see better results 🤷‍♂️

mdhaber replied:
2-opt is a little different than all possible permutations of the LHS. It is a greedy version of that idea. In the process of making all possible exchanges between two rows for the same column, you don't generate all possible Latin Hypercubes. But yes, best of e.g. 1000 randomly generated Latin Hypercubes and best of all possible Latin Hypercubes (if the problem is small) are also good ideas.

tupui replied:
In the first version I also had a mode with a simple loop to generate n LHS and take the best. At the time we discarded this as we said it was too simple and anyone could do this without problem. This could be part of an example to show how `discrepancy` can be useful.

tupui replied:
I went again through the main article: `Jin et al., "An efficient algorithm for constructing optimal design
       of computer experiments."
       Journal of Statistical Planning and Inference, 2005.`. Actually I should have something along what you proposed here.

There are 2 loops. The outer loop is the annealing part and the inner loop consist of doing random permutations and updating the best sample (similar to what you do). The two loops are linked through the temperature, it is used to select or discard a new sample. I think the current implementation of `dual_annealing` does not give access to the `temperature`. Maybe I could add a flag in order to allow passing it to the function (I think here https://github.com/scipy/scipy/blob/374e5bfec63e530c34514a08d9311352138182f3/scipy/optimize/_dual_annealing.py#L297).

cc @andyfaff

I updated the algorithm to make it a bit closer to the paper (without the temperature link for now). And it performs better (using your tests), especially for large dimensions.

In the end, if I manage to replicate the algorithm with the temperature and such, I would remove the `method` as it would not really make sense.

tupui replied:
I implemented the algorithm from the article. It seems good. I still need to check the performance agains the naive implementation and tests your `two_opt`. But the before and after is quite different now. For a design of size (10, 400), `np.where(before != after)` gives 1548 changes.

tupui replied:
@mdhaber for reference, I updated the description of the PR with the flowchart of the algorithm.

mdhaber replied:
What are the results of your tests of its effectiveness in a given amount of time (or time rate of discrepancy reduction)?

tupui replied:
If we reuse the best design and iterate like this, we obtain roughly the same performance with the new ESE. 

```python
original: 0.001231782193903852
basinhopping: 0.0011975526751670795
basinhopping: 0.0011950900444532842
...
basinhopping: 0.0011950900444532842
basinhopping: 0.0011950900444532842
basinhopping: 0.0011950900444532842
basinhopping: 0.0011950900444532842
swapper: 0.0003416202432611204
swapper: 0.0002696987661519046
...
swapper: 0.0001945242170817707
swapper: 0.00019050135380038746
ESE: 0.00025261993981051667
ESE: 0.00020622234542688744
```

But if we just do one shot with the same number of iterations, I have better results with ESE.
```python
original: 0.001231782193903852
basinhopping: 0.0011476753783927762
swapper: 0.0002172170249121752
ESE: 0.00018944240846363591
```

tupui replied:
For sure the previous strategy was just not working. Now the new one seem to be doing something better. Still maybe I have a bug in my implementation of the paper (like I was not entirely sure about the sample to give to the inner loop).

FYI, to get the results above I just took a previous commit (a72903b873d2831decafb3955f4205db13e5b0af) and copied the old `OptimalDesign` so that I had the 2 classes available.

mdhaber replied:
Are there are results in the paper you can try to reproduce that might hint at whether it's implemented correctly?
Also, it would be good to have stronger evidence that the method worth the added complexity. For instance, plots of discrepancy as a function of _time_ (as iterations could take different amounts of time) for a few different sizes/dimensions would make the decision easier.

tupui replied:
I will have a look at both, thanks Matt 😃

tupui replied:
I had a look at some numerical results from `Damblin et al., "Numerical studies of space filling designs: optimization of Latin Hypercube Samples and subprojection properties." Journal of Simulation, 2013.`. I could only check the convergence for C2 (canonical-subspaces: I only selected one subspace here) and minimum spanning tree (I had this code already). For the other figures (same with the paper from Jin2015) I would need to code another criterion: `phiP` and allow to change the metric used during the optimization.

The results seem to follow the findings from figures 9 and 10 of the paper.

Side note on the metrics: I am thinking of doing a separate PR to add MST and other distance variantes. What do you think?

<details>
  <summary>Reproducible code.</summary>

```python
import numpy as np
from scipy.stats import qmc
from scipy.spatial.distance import cdist
from scipy.sparse.csgraph import minimum_spanning_tree
import matplotlib.pyplot as plt


def l1_norm(sample):
    l1 = cdist(sample, sample, 'cityblock')
    return np.min(l1[l1.nonzero()])


# mindist
def l2_norm(sample):
    l2 = cdist(sample, sample)
    return np.min(l2[l2.nonzero()])


def linf_norm(sample):
    linf = cdist(sample, sample, 'chebyshev')
    return np.min(linf[linf.nonzero()])


def mst(sample, plot=False):
    """Minimum Spanning Tree criterion.

    MST is used here as a discrepancy criterion.
    Comparing two different designs: the higher the mean,
    the better the design is in terms of space filling.

    Parameters
    ----------
    sample : array_like (n, d)
        The sample to compute the discrepancy from.
    Returns
    -------
    mean, var : float
        Mean and standard deviation of the MST.
    """
    sample = np.asarray(sample)

    dist = cdist(sample, sample)
    mst = minimum_spanning_tree(dist)

    if (sample.shape[1] == 2) and plot:

        edges = np.where(mst.toarray() > 0)
        edges = np.array(edges).T

        fig, ax = plt.subplots()
        for edge in edges:
            ax.plot(sample[edge, 0], sample[edge, 1], c='k')
        ax.scatter(sample[:, 0], sample[:, 1])

        plt.show()

    return mst.data.mean(), mst.data.std()


qmc_gen = qmc.LatinHypercube(2)
doe = qmc_gen.random(20)
mst(doe, plot=True)


# figure 10: MST mean  0.7->0.8 ; std. 0.082 -> 0.065
n, d = 100, 10
qmc_gen = qmc.LatinHypercube(d)
doe = qmc_gen.random(n)
print(mst(doe))

for i in range(10):
    lhs_gen = qmc.OptimalLatinHypercube(d, doe, niter=1)
    doe = lhs_gen.random(n)
    print(mst(doe))


# figure 11: c2 disc 0.017 non optim and <0.01 optim
# C2 here need ** 0.5, it's a known difference due to the
# interpretation of the term C2 which needs or not to be squared.
# Our implementation is consistent with numerical results from the base literature.
n, d = 100, 10
qmc_gen = qmc.LatinHypercube(d)
doe = qmc_gen.random(n)
print(qmc.discrepancy(doe[:, :2]) ** 0.5)

for i in range(5):
    lhs_gen = qmc.OptimalLatinHypercube(d, doe, niter=1)
    doe = lhs_gen.random(n)
    # print(l2_norm(doe), l1_norm(doe), linf_norm(doe), qmc.discrepancy(doe[:, :2]) ** 0.5)
    print(qmc.discrepancy(doe[:, :2]) ** 0.5)
```
</details>

mdhaber replied:
> Side note on the metrics: I am thinking of doing a separate PR to add MST and other distance variantes. What do you think?

Interesting metric! I don't have any experience with it, but sounds useful.

As for the results, yes, there is some evidence that the match the paper. (Makes me wonder how simple random permutations would fare, but I suppose I can check that on my own when I am looking at this again.)

Did you happen to look into discrepancy as a function of time, or still working on that?

tupui replied:
> Interesting metric! I don't have any experience with it, but sounds useful.

I will put this on my list then, thanks.

> As for the results, yes, there is some evidence that the match the paper. (Makes me wonder how simple random permutations would fare, but I suppose I can check that on my own when I am looking at this again.)
> 
> Did you happen to look into discrepancy as a function of time, or still working on that?

Not yet, just need to find a bit of time for it 😅   But I also would like to be certain that I did not miss anything here.

tupui replied:
Ok here is the plot with respect to time, and yes the not fancy method works better... I am sending an email to the authors of these papers to see if I have done something wrong. Otherwise I will just go with your idea. Later on we can add a `method` param if somehow someone finds it 😅

![olhs_dummy](https://user-images.githubusercontent.com/23188539/117413563-685d0500-af16-11eb-874a-1dfea1e03976.png)

tupui replied:
@jschueller if you have some time maybe you can help here 😃  Maybe I have a very stupid bug here like mixing up some variables??

(Just careful about licensing, don't speak about how it's coded in OT)

tupui replied:
I just got one reply from G. Damblin and it seems like the algorithm is very difficult to parametrize. He is saying that it should be better in high dimension, but was not that convincing to me. So without further help and dedicated time to research this, I think that this approach will not work for us...

What I propose now is to use your `swapper` @mdhaber and if someone manages to code ESE and show a real gain (and I am now skeptical about this after having tried... thanks for all your questions which lead me to test this more by the way 😃 ) we can revisit.

(I would not try to push this before the release so we don't need to stress).

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -883,6 +977,177 @@ def reset(self):
+
+    def __init__(self, d, start_design=None, niter=1, method=None, seed=None):
+        super().__init__(d=d, seed=seed)
+        self.start_design = start_design
```
Ensure that it's an array, and we want to copy it. Go ahead and save it into `self.best_doe` initially rather than having two attributes, `best_doe` and `start_design`.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -846,6 +940,208 @@ def reset(self):
+                idx = np.argmin(best_disc)
+                best_disc = best_disc[idx]
+
+                if (best_disc - disc) < (th * self.rng.random()):
```
```suggestion
                if (best_disc - disc) < (th * self.rng.uniform()):
```

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -209,39 +211,56 @@ def test_update_discrepancy(self):
             update_discrepancy(x_new, space_1[:-1], disc_init)
 
+    def test_perm_discrepancy(self):
+        qmc_gen = qmc.LatinHypercube(5)
```
This is causing the other test failure in the custom `check_random_state`. Please check the documentation of that one. It says it will turn `int`s into a `Generator` but instead it raises this error.
```
E   ValueError: NumPy 1.16 doesn't have Generator, use either NumPy >= 1.17 or `seed=np.random.RandomState(seed)`
```

tupui replied:
Indeed, thanks. It was missing explicit seeding to work. It is by design that having `seed=None` only calls `np.random.Generator`. We dit this intentionally so that `None` would give the recommended way only. The documentation is also not talking about `RandomState` on purpose.

tupui commented:
@mbaudin47 this might interest you 😉

mdhaber commented:
@tupui is this ready for another look? I was waiting for the last box above to be checked and a test of the effectiveness relative to random swaps.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -940,56 +940,61 @@ def reset(self):
+    The specified optimization `method` is used to select a new set of
+    permutations to perform. If `method` is None,
+    `~scipy.optimize.dual_annealing` optimization is used. `niter` set of
+    optimizations are performed.
```
I thought you removed `method` above?

tupui replied:
Yes, yes, as I said I did not touch yet the docstrings too much as I am rewriting things.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -940,56 +940,61 @@ def reset(self):
     >>> sample = sampler.random(n=5)
     >>> sample
-    array([[0.84052691, 0.53664833],
+    array([[0.84052691, 0.53664833],  # random
```
Aren't we using `default_rng()` which, under the hood, gets an explicit seed so that it's not random?

tupui replied:
Same

tupui commented:
@mdhaber I refactored to only keep your swapping method. Let me know what you think.

The unknown I had was about making a generic function which would allow another type of `QMCEngine`, but then there might be little value in such function as it would just be: generate n design and take the one with the best metric.
As for `OptimalLatinHypercube`, if there is a future interest, we could easily at a `method` to change the metric. The value here comes from the fancy swapping and the faster update of the metric (there are expressions for some other discrepancies). For now, I think this would be enough.

mdhaber commented:
Thanks @tupui. I will try to take a look before Wednesday, but I'm working on the requested `bootstrap` enhancements.

tupui commented:
> Thanks @tupui. I will try to take a look before Wednesday. Since you said this does not need to go in this release, though, I'm working on the requested `bootstrap` enhancements first.

Sure it's not a priority and I don't mind if we merge this later. Thanks Matt.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+        Initial sample to optimize. `LatinHypercube`
+        is used to generate a first design otherwise.
+    n_perturbations : int, optional
+        Number of perturbation to perform. Default is 10000.
```
```suggestion
        Number of perturbations to perform. Default is 10000.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+    n_perturbations : int, optional
+        Number of perturbation to perform. Default is 10000.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
```
```suggestion
        If `seed` is None, the `numpy.random.Generator` singleton is used.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
```
```suggestion
        If `seed` is already a ``Generator`` instance, then that instance is
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+
+    Notes
+    -----
+    `n_perturbations` set of random permutations are performed aginst a LHS
+    sample. The best design based on the centered discrepancy is constantly
+    update. Centered discrepancy-based design shows better space filling
+    robustness toward 2D and 3D subprojections. Distance-based design shows
+    better space filling but less robustness to subprojections [2]_.
```
```
A sequence of random perturbations are performed on an LHS sample.
Each perturbation consists of swapping a random coordinate of a randomly-chosen
sample point with the corresponding coordinate of another randomly-chosen sample
point, preserving the Latin Hypercube properties of the sample.
Perturbations that reduce the centered discrepancy of the sample are preserved;
perturbations that increase the centered discrepancy are rejected. After
`n_perturbations` are considered, the process terminates, resulting in a new LHS
with substantially lower centered discrepancy.
```
I'm not sure if the notes about centered discrepancy are needed here. It seems like it's being compared to "distance-based design", but there is no option to optimize a distance-based design, so why compare the two here? 
> Centered discrepancy-based design shows better space..

Better compared to what?

> Distance-based design

Is this an alternative to centered discrepancy?

tupui replied:
My intent was to say that the centered discrepancy was a good choice and on purpose. You could chose to optimize the wrap around discrepancy, but from the linked article they have seen worse performance. To have the nice update function, you also need to derive it. It has been done for not all measures (and I just coded and fixed this one).

Here we don't have distance-based measures (I have an open issue for that but very easy so for now I leave it for Sprints...). Distance measures are simpler things like L-inf or mean distance between points, etc.

tupui replied:
I simplified the doc here.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
 
 
+class OptimalLatinHypercube(QMCEngine):
+    """Optimal Latin hypercube sampling (OLHS).
```
```suggestion
    """Greedy optimization of a Latin hypercube sample (OLHS).
```
I think we should mention it's "greedy" somewhere, and I would avoid stating that it's "optimal" because it's not even guaranteed to be locally optimal.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+
+    """
+
+    def __init__(self, d: IntNumber, *, start_sample: npt.ArrayLike = None,
+                 n_perturbations: IntNumber = 10000,
+                 seed: SeedType = None):
```
Is there supposed to be a space around the equals sign when used with types? 
(Do I understand that type information can be moved to a separate file to avoid cluttering the code?)

tupui replied:
I know it looks strange but yes there is.

Also yes, the types could be moved to a stub file (same name with a different extension `pyi`). It was actually the case in the first version, and then we said that it would be better to not duplicate all the files like this and have the types inline.

mdhaber commented on scipy/stats/qmc.py:
```diff
@@ -23,6 +23,7 @@
    Sobol
    Halton
    LatinHypercube
+   OptimalLatinHypercube
```
Need to remove `_qmc.pyi` above, I think?

tupui replied:
Totally! A left over from a merge I guess.

tupui replied:
Removed.

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -209,39 +212,57 @@ def test_update_discrepancy(self):
             disc1 = np.sum(np.prod((1
-                                    + 1/2*np.abs(xij-0.5)
-                                    - 1/2*np.abs(xij-0.5)**2), axis=1))
+                                    + 1 / 2 * np.abs(xij - 0.5)
+                                    - 1 / 2 * np.abs(xij - 0.5) ** 2), axis=1))
```
Let's not do this in this PR. They are subjective, and they are unrelated to rest of the PR.

tupui replied:
Ok sure I will remove all these.

tupui reviewed (commented):
Thanks for the review Matt!

(I will try to get some time this WE to have a look at the decorator and the moment PR. Still catching up 😅)

tupui reviewed (commented):
I think I went through all your comments. Let me know if there is anything else.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+        self.num_generated += n
+        return best_sample
+
+    def reset(self):
```
`num_generated` is private, and it's not part of the algorithm, so if no `start_sample` is provided, `reset` has no meaningful effect, right? 

Er, `num_generated` is essentially undocumented... it probably _should_ be private and marked as such with a preceding underscore. 

I see it mentioned in the [`QMCEngine`](https://33172-1460385-gh.circle-artifacts.com/0/html-scipyorg/reference/generated/scipy.stats.qmc.QMCEngine.html#scipy.stats.qmc.QMCEngine) documentation. This reminds me of gh-13225. It can be a headache when there are pseudo-public parameters that users start relying on.

tupui replied:
> `num_generated` is private, and it's not part of the algorithm, so if no `start_sample` is provided, `reset` has no meaningful effect, right?

True! I can remove this function and put the `self.lhs = LatinHypercube(self.d, seed=self.rng)` directly in `self.random`.

> Er, `num_generated` is essentially undocumented... it probably _should_ be private and marked as such with a preceding underscore.
> 
> I see it mentioned in the [`QMCEngine`](https://33172-1460385-gh.circle-artifacts.com/0/html-scipyorg/reference/generated/scipy.stats.qmc.QMCEngine.html#scipy.stats.qmc.QMCEngine) documentation. This reminds me of [gh-13225](https://github.com/scipy/scipy/issues/13225). It can be a headache when there are pseudo-public parameters that users start relying on.

Yes it is in the doc. I kind off wanted to have it as most of these methods are iterative and it makes sense to be able to access the state.

LHS is particular in a sense that it is a one shot algorithm. You give n and you are done with it. But there are methods in the literature to extend make LHS iterative. I never tried to code anything here though and did not find any public code. So I don't know how it performs.

Shall I add a note in the doc to warn about this attribute is public and we should be careful not to remove it? (or you prefer the other way around? private and we warn (or not) not to use it?)

mdhaber replied:
My opinion would be to either go all in and document it as public or change it to `_num_generated` for 1.7.1 before there has been much time for people to notice and start relying on it. I suppose I would prefer making it private until we can think of good reasons for exposing it.

tupui replied:
I propose to do this in a follow up and change this in all classes if still wanted.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+        return best_sample
+
+    def reset(self):
+        """Reset the engine to base state.
```
And if `start_sample` is provided, then `reset` does a shallow reset of the object to the initial state - shallow because it does not reset the underlying RNG unless an `int` is provided for `seed`. Correct? 

I'm thinking about how it can be tested. I guess don't see a real use case of the method unless `seed` is an `int`, so the test would just make sure that `reset` and a new object (constructed with a deep-copied RNG) do the same things.

tupui replied:
I removed this reset as per the other comment. It's not needed :)

mdhaber replied:
It is still inheriting the `reset` method from `QMCEngine`, but at least that creates a deep copy.

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -534,6 +555,58 @@ def test_sample_stratified(self):
+    def test_continuing(self, *args):
+        pytest.skip("Not applicable: not a sequence.")
+
+    def test_fast_forward(self, *args):
+        pytest.skip("Not applicable: not a sequence.")
+
```
If it has the method, I think there should be a test. Same with `reset`.

tupui replied:
What about if for cases like this (LHS and this one), I return a warning saying that it's not a sequence and that there is no correlation/link between the previous and next calls?

mdhaber replied:
Well, you could, but it would seem even more to me that something is wrong, either with `QMCEngine` insisting on having a `fast_forward` method or with this class being a subclass of `QMCEngine`.

Setting that question aside, you could pretend that the `fast_forward` was useful and override the test with something like this:
```
    def test_fast_forward(self, *args):
        lhs = qmc.LatinHypercube(d=2, seed=95123684)
        lhs.fast_forward(10)
        sample1 = lhs.random(10)

        lhs.reset()
        lhs.random(10)
        sample2 = lhs.random(10)
        assert_allclose(sample1, sample2)
```

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -534,6 +555,58 @@ def test_sample_stratified(self):
+        sample_ref = lhs.random(n=20)
+        disc_ref = qmc.discrepancy(sample_ref)
+
+        # all defaults
```
This doesn't seem to use all defaults. Do you mean e.g. `# using a random initial LHS`?

tupui replied:
I updated the tests. So closing this one as not relevant anymore.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+        else:
+            self.start_disc = np.inf
+
+        self.lhs = LatinHypercube(self.d, seed=self.rng)
```
This isn't used if there is a `start_sample`, right?

For this and other reasons - e.g. the `n` parameter of `random` has no effect if a `start_sample` are provided - it seems that `start_sample` really throws a wrench into things. I'll reserve judgement for now, but it's hinting that what we really want to have is a function that optimizes a Latin hypercube sample. Maybe it's fine to also have an `OptimalLatinHypercube` class for convenience, but even though intuitively it seems OK for such a class to accept a `start_sample`, it turns out to have some problems.

tupui replied:
I moved this in `random` as it's only used there and we have no need for it before. When resetting, the same seed will go there, hence the same sampling.

I did not think about `n`... then I agree that this looks a bit weird to have a non-used parameter. So what you are saying is that instead of having this class, I could just refactor `random` to be a standalone function to optimize LHS? This would look the best I guess as it's a one time optimization and we do not need to have an object for that.

Because in the end the convenient class should accept any `QMCEngine` and use `update_discrepancy`. But the interface would be complicated as we would need to allow other cost functions and update functions. We could be re-writting some ersatz of `optimize` 😅 

I will wait for your input here I guess. pinging you here for conveniency @mdhaber

mdhaber replied:
Another possibility would be for `LatinHypercube` to return a `LatinHypercubeSample` object, which would essentially just be a `np.ndarray`, but it would have an `optimize` method. (Or it could be an object that holds the array as an attribute, and it has an optimize method). Not sure how SciPythonic this would be, but we are starting to do that sort of thing in `scipy.stats` with result objects that have methods for generating a confidence interval.

tupui replied:
Nice idea! Do you think that doing so could bring a bit of inconsistency as the other classes return plain `np.ndarray`? Although it could paves the way towards adding new features to the other engines (I still have in mind nice `repr` (optional with plots) and having an object could help, etc.).

What about another method (or `classmethod`) in `LatinHypercube`? This would make it more discoverable.

mdhaber replied:
I suppose they can all be `scale`d.

tupui replied:
All right then I will augment the outputs of the `QMCEngines` and make `random(self, n: int) -> QMCSample` with `QMCSample(np.ndarray)`:
* `discrepancy` property with an optional parameter for the method
* `scale` could scale inplace or not
* (`__repr__` not sure)
* and in this case `optimize`, same can be inplace (should be I guess)

cc @rgommers as even if the submodule is new, this might be surprising for some user and I am not sure if we need/should warn the users before doing this. Or you may have another idea we should explore.

rgommers replied:
I read the code, but kind of missed the actual problem (my brain may not be working anymore). `random()` just returns `best_sample` in all cases it looks like (`start_sample` is also an `(n, d)`-shaped array). Can you spell it out for me?

tupui replied:
Yes both are `(n,d)`-shaped `np.ndarrays`.

In the end, we can see that `OptimalLatinHypercube` would not really benefit from being a class. Everything is in this `random` function. The feature could either: be a single function; be included in `LatinHypercube` with a new (class?)method; or `random` could return a subclass of `np.ndarray` which will have more functionalities. Doing so, we could harmonise the other engines and we would return a common `QMCSample(np.ndarray)` which would have things like discrepancy and scaling.

rgommers replied:
Ah makes sense. Not a fan of returning an `ndarray` subclass, because subclasses easily get lost due to the `asarray` usage everywhere. The other two options sound better.

tupui replied:
Ok thanks. I gave this more though and can see these three options to continue (sorry for this, I want to make this right as this is a new module):

1. Add in the class `LatinHypercube` the method `random_optimized(n, *, n_perturbations)`.
I am not sure it's a good idea to allow `n` and `start_sample`. As Matt said, this would make `n` useless if we have `start_sample`. So we might be just using `n` for both. Also maybe keeping this name is better to have the same interface for everyone.

```python
engine = LatinHypercube(d=2)
sample = engine.random_optimized(100)

start_sample = ...
sample = engine.random_optimized(start_sample)
```

2. Just add a parameter `optimize` in `LatinHypercube`.

```python
engine = LatinHypercube(d=2, optimize=True, n_perturbations=10000, start_sample=None)
sample = engine.random(100)

start_sample = ...
engine = LatinHypercube(d=2, optimize=True, n_perturbations=10000, start_sample=start_sample)
sample = engine.random(100)  # but this is a problem as useless param
```

With both of these options, there is the parameter `centered` which would be useless.

3. Going with the function only, advantage: no conflict with `centered` being not used.
```python
sample = latin_hypercube_optimized(d=2, n=100)

sample = latin_hypercube_optimized(d=2, n=start_sample)
```

rgommers replied:
Looking ahead: if there's a change there may be multiple "optimization" strategies, it'd be better as a string rather than a boolean keyword. Not 100% sure that's the case here, but we've made the boolean keyword mistake way too often already.

tupuinui replied:
Good point, thanks! Am I reading that you prefer option 2 @rgommers?

tupui replied:
Good point, thanks! Am I reading that you prefer option 2 @rgommers?

rgommers replied:
Not really, I'm just skimming the discussion - too many other things going on right now. This thing just jumped out at me.

tupui replied:
@mdhaber do you have any preference? On my side, I think I like 2. And we can also not include `start_sample` for now and see if some people need it.

mdhaber replied:
I think the much more common use case will be for someone to want to generate an optimize Latin hypercube - much more often than to optimize an existing Latin hypercube - so I don't think we really need that capability.

With that in mind, adding an argument to the existing method that generates a Latin hypercube (`random`) seems simpler than creating a new function or method. So yes, I think option 2 is the best above.

I agree that we should not add `start_sample` or similar at this time (because of the first point).

I think that `optimize` and `n_perturbations` are redundant for now, if we only have one optimization method. If `n_perturbations` is not 0, then we know the user wants to optimize. We could add an argument to select the optimization method if/when we add another method.

On the other hand, `n_perturbations` might not be an appropriate parameter name for a different optimization method. So an alternative would be to keep only the `optimize` keyword, instead... maybe call it `optimization` so that it's more natural for the value to be the name of the optimization method. In that case, we can use a heuristic to detect convergence rather than having the user specifying the number of permutations. We could add options for control in the future if it is a requested feature and if we have a better idea of what parameters would be common to all optimization methods.

tupui replied:
Thanks Matt! I will refactor to add the parameter and some stopping logic.

tupui replied:
Done @mdhaber, let me know what you think 😃

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+            col = rng_integers(self.rng, *bounds[0])
+            row_1 = rng_integers(self.rng, *bounds[1])
+            row_2 = rng_integers(self.rng, *bounds[2])
+            doe = best_sample.copy()
```
If `_perturb_discrepancy` just predicts the new discrepancy (without mutating its first argument), is this needed? A copy of `start_disc` should be made once at the beginning, but not every iteration. 

(If `_perturb_discrepancy` _does_ mutate the array, it should probably return the mutated array, or it should perform the copy itself.)

tupui replied:
Totally! Maybe a leftover of the previous version...

mdhaber reviewed (commented):
The implementation looks correct, but this needs some input validation and additional tests, and I think there is an interface issue (whether `start_sample` is appropriate for an `OptimalLatinHypercube` _object_ or whether we really want to have a function that optimizes Latin hypercube samples) to think about.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+    """Greedy optimization of a Latin hypercube sample (OLHS).
+
+    Optimize the sample by doing random permutations of coordinates to lower
+    the centered discrepancy [1]_. The resulting sample is still a LHS.
```
Yes the equation is in there. It is true that the article refers to the fancy algorithm. I can put a more general one if you want. Although I am kind of hopping that interested people to actually go read this would find a way to do it right?? 😅

mdhaber replied:
K, thought it might have been left over from before. Yes, the algorithm is not very complicated : )

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -950,18 +968,56 @@ class LatinHypercube(QMCEngine):
            [6.80338249, 3.08795949],
            [2.65448791, 3.83491828]])
 
+    Alternatively, optimized LHS can be constructed with `optimization`.
```
This is not an alternative to scaling, so it should either go before scaling or the word "Alternatively" should be changed or dropped.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -950,18 +968,56 @@ class LatinHypercube(QMCEngine):
            [2.65448791, 3.83491828]])
 
+    Alternatively, optimized LHS can be constructed with `optimization`.
+    It should give a sample with a lower discrepancy at a higher computational
```
"should" is unnecessary here, IMO. Or, this can all be:

> Use the `optimized` keyword argument to produce an LHS with
lower discrepancy at higher computational cost.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -860,18 +968,56 @@ class LatinHypercube(QMCEngine):
         super().__init__(d=d, seed=seed)
         self.centered = centered
 
+        optimization_methods: Dict[str, Callable] = {
+            "random-CD": self._random_optimal,
+        }
+
+        if optimization is None:
+            self.optimization = self._random
+        elif optimization in optimization_methods:
+            self.optimization = optimization_methods[optimization]
+        else:
+            raise ValueError(f"{optimization!r} is not a valid optimization"
```
```suggestion
        optimization_methods: Dict[str, Callable] = {
            None: self._random,
            "random-CD": self._random_optimal,
        }

        try:
            self.optimization = optimization_methods[optimization]
        except KeyError:
            raise ValueError(f"{optimization!r} is not a valid optimization"
```

tupui replied:
Better yes 🙌 I will just add `raise from` to have a better trace back.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -860,18 +968,56 @@ class LatinHypercube(QMCEngine):
+            LHS sample.
+
+        """
+        return self.optimization(n)
```
This is subjective, but I don't think it makes sense for `random` to always call `self.optimization`, which may just be `self._random`. Seems to me that `random` should produce a random sample (e.g. call `self._random`) and then, if `self.optimization` is not None, call the appropriate optimization method.

tupui replied:
Agreed. I wanted to avoid too much logic here. Alternatively I could replace the attribute's name with something like `lhs_method`. This way we still have a clear separation of both algorithms.

mdhaber replied:
You can keep the separation of you want. It could literally be just:
```
sample = self.random(n)
if self.optimization is not None:
    sample = self.optimization(sample)
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +1044,63 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+
+        n_nochange = 0
+        n_iters = 0
+        while "Optimization loop":
```
Maybe refactor this in terms of `while n_nochange < 10 and n_iters < 10000` to avoid the appearance of an infinite loop and remove some lines below.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -898,6 +1044,63 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+                n_nochange = 0
+            else:
+                # stopping criterion: wait for no successive improvement
+                if abs(best_disc - disc) < 1e-10:
```
I'm not sure if `abs` is appropriate here. If `disc` is much greater than `best_disc`, it seems to me that `n_nochange` should increase. Basically, I don't think you need this additional `if` statement at all. If `disc < best_disc` above is False, then there is no change and `n_nochange` should increase. Right? Isn't the idea to just stop if no _improvement_ is made in some threshold number of iterations?

tupui replied:
Ok sure I will remove it. It should then be more than 10 (it's arbitrary in any case...) as the changes being totally random we might just have bad luck for a few iterations.

As you said, we will see what people want want they start to use it.

mdhaber replied:
Well maybe you had a different stopping criterion in mind, if it was intentional. You're not just looking for a local minimum but instead a local _plateau_ where it's hard to find permutations that bring the discrepancy up _or_ down?

tupui replied:
I have revisited this removing the abs as you suggested. I also put private attributes to control the params if someone really needs to and cannot wait a release.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -901,6 +903,17 @@ class LatinHypercube(QMCEngine):
         Dimension of the parameter space.
     centered : bool, optional
         Center the point within the multi-dimensional grid. Default is False.
+    optimization : {None, "random-CD"}, optional, default None
```
```suggestion
    optimization : {None, "random-CD"}
```
> When a parameter can only assume one of a fixed set of values, those values can be listed in braces, with the default appearing first:
```
order : {'C', 'F', 'A'}
    Description of `order`.
```
See https://numpydoc.readthedocs.io/en/latest/format.html#parameters. 

Besides that, it should be _either_ `optional` with the default described in the text OR `default: None`, not both.

tupui replied:
Thanks I was not sure about the syntax here. I might put it in the text then to match the style of the file.

mdhaber reviewed (changes requested):
I think the interface is a lot better. If users use this feature, they will probably complain about wanting control over the optimization process. I think that's a _good_ thing because it will give us a reason to develop this further. But for now, this is much better than most of the alternatives we've considered because it doesn't change the existing interface in a way that we're unsure about. I think it's pretty harmless to add this one option, which seems very reasonable to have regardless of where things go. It's much better than adding an entire class that doesn't really fit the `QMCEngine` model or even a standalone function which would initially have only one very restricted purpose (to optimize the output of `LatinHypercube.random`).
Suggested some simplifications to the documentation and code, but looking close.

mdhaber reviewed (commented):
Looks good.

Need to fix the rendering of the new example.

![image](https://user-images.githubusercontent.com/6570539/127556750-8ebafda9-dbf8-478b-bb8d-a25efe56e1c3.png)

I'd also prefer that the optimization option be case-insensitive. I accidentally tried `'random-cd'` and got an error. That annoys me, personally, but I know there are different opinions about that sort of thing.

Here's a simple script to see how the optimization does:
```
import numpy as np
from scipy.stats import qmc

d, n = 3, 100

rng = np.random.default_rng(9842398122)
lhs_seeds = rng.integers(1000, size=1000)

for seed in lhs_seeds:
    qmc_gen = qmc.LatinHypercube(d, seed=seed)
    doe = qmc_gen.random(n)
    disc = qmc.discrepancy(doe)
    
    qmc_gen = qmc.LatinHypercube(d, seed=seed, optimization='random-CD')
    doe = qmc_gen.random(n)
    optdisc = qmc.discrepancy(doe)
   
    print(optdisc/disc)
```

It would be nice @tupui if you could produce a few plots showing the discrepancy as a function of iteration so we can gauge whether the convergence criteria are about right. But that is make-better, so we could wait to consider it once we get requests to do so.

But the documentation is OK and the tests seem to check what they need to: the discrepancy update calculation works, the method argument produces a reasonable error message when it's not valid, the optimized design is better than the original, and it's still a Latin Hypercube. So not too much holding this up.

tupui commented:
> Looks good.
> 
> Need to fix the rendering of the new example.

Ooops, should be fixed now.

> It would be nice @tupui if you could produce a few plots showing the discrepancy as a function of iteration so we can gauge whether the convergence criteria are about right. But that is make-better, so we could wait to consider it once we get requests to do so.

Ok sure, I will try to make something useful.

> But the documentation is OK and the tests seem to check what they need to: the discrepancy update calculation works, the method argument produces a reasonable error message when it's not valid, the optimized design is better than the original, and it's still a Latin Hypercube. So not too much holding this up.

Glad to read, thanks Matt 🥳

tupui commented:
@mdhaber I have some results. Seems like I was lucky and 100, 10_000 seems fine.

Note that the values are sometimes not round (ex. 9830) because I am doing the mean over some realizations (here 5 as it does not change much).
```
d=10, Ns=200

Default values: 10000, 100 | 0.4153428604265345 [  55 9830]
Case 1
1000, 100 | 0.5698480676322689 [   2 1000]
5000, 100 | 0.4499603262234236 [  37 5000]
10000, 100 | 0.4153428604265345 [  55 9830]
20000, 100 | 0.41132065360601955 [  100 10716]
50000, 100 | 0.41132065360601955 [  100 10716]
Case 2
10000, 10 | 0.7494520426748597 [ 10 297]
10000, 100 | 0.4153428604265345 [  55 9830]
10000, 1000 | 0.4144494432252772 [   17 10000]
10000, 2000 | 0.4144494432252772 [   17 10000]
Case 3
50000, 10 | 0.7494520426748597 [ 10 297]
50000, 100 | 0.41132065360601955 [  100 10716]
50000, 1000 | 0.35679136503433057 [   71 50000]
50000, 2000 | 0.35679136503433057 [   71 50000]

-----------------------------------------------------------------

d=5, Ns=100

Default values: 10000, 100 | 0.3119049788602787 [ 100 3947]
Case 1
1000, 100 | 0.3815846896444995 [  10 1000]
5000, 100 | 0.3119049788602787 [ 100 3947]
10000, 100 | 0.3119049788602787 [ 100 3947]
20000, 100 | 0.3119049788602787 [ 100 3947]
50000, 100 | 0.3119049788602787 [ 100 3947]
Case 2
10000, 10 | 0.60966387214425 [ 10 114]
10000, 100 | 0.3119049788602787 [ 100 3947]
10000, 1000 | 0.2778197429019117 [   23 10000]
10000, 2000 | 0.2778197429019117 [   23 10000]
Case 3
50000, 10 | 0.60966387214425 [ 10 114]
50000, 100 | 0.3119049788602787 [ 100 3947]
50000, 1000 | 0.24349852818152812 [  259 48654]
50000, 2000 | 0.24292300734518152 [   72 50000]
```


<details> 
  <summary>Code to reproduce </summary>

To reproduce you also need to add `self.debug = [n_nochange, n_iters]` before `_random_cd` returns.

```python
import numpy as np
from scipy.stats import qmc

d, n = 10, 200
n_conv = 5

rng = np.random.default_rng(9842398122)
lhs_seeds = rng.integers(1000, size=n_conv)


def opt_disc(seed, n_nochange, n_iters):
    qmc_gen = qmc.LatinHypercube(d, seed=seed)
    doe = qmc_gen.random(n)
    disc = qmc.discrepancy(doe)

    qmc_gen = qmc.LatinHypercube(d, seed=seed, optimization='random-cd')
    qmc_gen._n_nochange = n_nochange
    qmc_gen._n_iters = n_iters
    doe = qmc_gen.random(n)
    optdisc = qmc.discrepancy(doe)

    return np.array([optdisc/disc, *qmc_gen.debug])


n_nochange = 100
n_iters = 10_000

data = [opt_disc(seed, n_nochange=n_nochange, n_iters=n_iters)
         for seed in lhs_seeds]
data = np.mean(data, axis=0)
ratio = data[0]
debug = data[1:].astype(int)
print(f"Default values: {n_iters}, {n_nochange} | {ratio} {debug}")

print("Case 1")

n_nochange = 100
for n_iters in [1000, 5_000, 10_000, 20_000, 50_000]:
    data = [opt_disc(seed, n_nochange=n_nochange, n_iters=n_iters)
            for seed in lhs_seeds]
    data = np.mean(data, axis=0)
    ratio = data[0]
    debug = data[1:].astype(int)
    print(f"{n_iters}, {n_nochange} | {ratio} {debug}")

print("Case 2")
n_iters = 10_000
for n_nochange in [10, 100, 1_000, 2_000]:
    data = [opt_disc(seed, n_nochange=n_nochange, n_iters=n_iters)
            for seed in lhs_seeds]
    data = np.mean(data, axis=0)
    ratio = data[0]
    debug = data[1:].astype(int)
    print(f"{n_iters}, {n_nochange} | {ratio} {debug}")

print("Case 3")
n_iters = 50_000
for n_nochange in [10, 100, 1_000, 2_000]:
    data = [opt_disc(seed, n_nochange=n_nochange, n_iters=n_iters)
            for seed in lhs_seeds]
    data = np.mean(data, axis=0)
    ratio = data[0]
    debug = data[1:].astype(int)
    print(f"{n_iters}, {n_nochange} | {ratio} {debug}")

```

</details>

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -860,15 +968,41 @@ class LatinHypercube(QMCEngine):
+        try:
+            if optimization is not None:
+                optimization = optimization.lower()
+            self.lhs_method = lhs_methods[optimization]
```
@BvB93 can you help here? I am not sure what's best to do...

`scipy/stats/_qmc.py:996: error: Incompatible types in assignment (expression has type "str", variable has type "Optional[Literal['random-cd']]")  [assignment]`

BvB93 replied:
I think this might be a use-case for a `# type: ignore[assignment]` comment.

In my experience mypy's support for string literals is not that great; 
thus you're getting an error as mypy thinks the `str.lower` call returns a normal string instead of the appropiate string-literal.

tupui replied:
Ok then I will use this. Thanks! It's quite annoying that it is not well supported...

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -1022,12 +1022,12 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
         """Base LHS algorithm."""
         if self.centered:
-            samples: np.ndarray | float = 0.5
+            samples = np.array(0.5)
```
Concerning to use 0d array to avoid an errant type complaint. Is this ever going to return a 0d array?

tupui replied:
I tested this and it does return an 0d array if you ask for 0 sample and 0 dim (also in the tests). I can revert if you prefer?

mdhaber replied:
The code
```python3
from scipy.stats.qmc import LatinHypercube
lhs = LatinHypercube(0, centered=True)
lhs.random(0)
```
does not return a 0d array; it returns a 2d array with shape (0, 0). I think that is the correct behavior. I was concerned with the possibility of this function returning a 0d array, which would not be correct.

Still, I am also not sure that we should change the way we write this code to satisfy mypy. Use of a Python float here is what you had in mind originally because it's more natural, and I think the rest of the SciPy codebase agrees.

tupui replied:
Sure, agreed. I just reverted this line.

tupui commented:
@mdhaber I think I covered all the points. There is just this 0d array question (either the current version is fine or I revert it back to a float) but otherwise all is green now.

tupui commented:
In case you did not see @mdhaber, Azure CI is not working right now due to #14529

mdhaber commented:
Yes, I noticed. Would you prefer to wait until those checks can run to merge, or do you think that changes since the last successful fun on Azure are unlikely to cause the Azure tests to fail?

tupui commented:
> Yes, I noticed. Would you prefer to wait until those checks can run to merge, or do you think that changes since the last successful fun on Azure are unlikely to cause the Azure tests to fail?

IMO you can merge. I did not change anything since the last full run on Azure that was failing the type.

mdhaber commented:
Ok! I think this is a valuable change to `LatinHypercube`. We added the new functionality (optimization of a Latin Hypercube sample to lower the centered discrepancy) with a minimal change to the interface (just the addition of an optional `optimization` keyword-only argument) to leave open the possibility of expanding in the future. The documentation changes seem complete, and the essential tests have been added: 

- accuracy of the method used to update the centered discrepancy w.r.t. elementary permutations of a Latin hypercube
- the optimized Latin hypercube is still a Latin hypercube
- the optimized Latin hypercube has lower centered discrepancy than the original
- basic input validation on the `optimization` keyword argument

Thanks @tupui!

tupui commented:
Thanks Matt for your extensive review which really helped to uncover a lot of things!
